### PR TITLE
Feature/gif video tag

### PR DESCRIPTION
**Pull request by selul**

### CODE REVIEW

File: inc/tag_replacer.php [hunk 1]
```diff
@@ -62,22 +62,97 @@ public function init() {
...
+	 * @param string $content The content to process.
+	 *
+	 * @return bool
+	 */
+	public function img_to_video( $image_url, $image_tag, &$content ) {
+		if ( $this->settings->get( 'img_to_video' ) === 'disabled' ) {
+			return false;
+		}
+
+		$file = get_headers( $image_url, 1 );
```
Comment (selul) [inline inc/tag_replacer.php]: we should avoid checking the content type in this way, checking the extension only is enough.

File: inc/tag_replacer.php [hunk 2]
```diff
@@ -62,22 +62,97 @@ public function init() {
...
+		$link_svg = apply_filters(
+			'optml_content_url',
+			$image_url,
+			array('width' => 'auto',
+				'height' => 'auto',
+				'format' => 'svg',
+			)
+		);
+
+		$video_tag = '<video autoplay muted loop playsinline poster="' . $link_svg . '">' .
```
Comment (selul) [inline inc/tag_replacer.php]: we should preserve classes and custom attributes on replace.

File: inc/url_replacer.php [hunk 3]
```diff
@@ -169,16 +172,22 @@ public function build_image_url(
...
-				'apply_watermark' => apply_filters( 'optml_apply_watermark_for', true, $url ),
-			]
-		);
+		$arguments = [
+			'signed'          => $this->settings->use_lazyload() ? false : $this->is_allowed_site,
+			'apply_watermark' => apply_filters( 'optml_apply_watermark_for', true, $url ),
+
+		];
+
+		if ( isset( $args['format'] ) && ! empty( $args['format'] ) ) {
```
Comment (selul) [inline inc/url_replacer.php]: it makes more sense to me to have the format in `$args` rather `$arguments`. I have imagined `$arguments` those more like global settings applied to all the images, even if the variable name is not that intuitive 🙉

File: inc/app_replacer.php [hunk 4]
```diff
@@ -337,13 +337,19 @@ public function can_replace_url( $url ) {
...
 		$url_parts = parse_url( $url );
 
 		if ( ! isset( $url_parts['host'] ) ) {
 			return false;
 		}
+		// https://mlynllm40tdp.i.optimole.com/HNUQD3PR0fo/w:auto/h:auto/q:auto/
+		// http://c37866cd.ngrok.io/wp-content/uploads/2019/07/68747470733a2f2f6d65646961312e7465
+		// 6e6f722e636f6d2f696d616765732f65616337663764383533346630383433656264373037313031623865663766
+		// 642f74656e6f722e6769663f6974656d69643d3131363939363038.gif
+		// $this->set_properties();
```
Comment (selul) [inline inc/app_replacer.php]: let's remove this comment

File: inc/app_replacer.php [hunk 5]
```diff
@@ -337,13 +337,19 @@ public function can_replace_url( $url ) {
...
 		if ( false === ( isset( $this->possible_sources[ $url_parts['host'] ] ) || isset( $this->allowed_sources[ $url_parts['host'] ] ) ) ) {
-			return false;
+			return false;                     // here
```
Comment (selul) [inline inc/app_replacer.php]: the comment seems redundant

**Comment by selul** [comment]
we should add here a sample PHPUnit test and also a cypress scenario for a testcase when the origin will have the GIF feature deployed.

File: inc/tag_replacer.php [hunk 6]
```diff
@@ -62,22 +62,106 @@ public function init() {
...
+		$video_tag = str_replace(
+			[
+				'src=',
+				'<img',
+				'/>',
+			],
+			[
+				'original-src=',
+				'<video autoplay muted loop playsinline poster="' . $link_png . '"',
+				'><source src="' . $link_mp4 . '" type="video/mp4"><source src="' . $link_webm . '" type="video/webm">sss</video',
```
Comment (selul) [inline inc/tag_replacer.php]: that `sss` seems redundant here

File: inc/tag_replacer.php [hunk 7]
```diff
@@ -62,22 +62,106 @@ public function init() {
...
+	public function img_to_video( $image_url, $image_tag, &$content ) {
+		if ( $this->settings->get( 'img_to_video' ) === 'disabled' ) {
+			return false;
+		}
+
+		if ( $this->is_valid_mimetype_from_url( $image_url, [ 'gif' => false ] ) ) {
+			return false;
+		}
+
+		$link_webp = apply_filters(
```
Comment (selul) [inline inc/tag_replacer.php]: We don't seem to use this anywhere.

File: inc/image.php [hunk 8]
```diff
@@ -107,13 +109,17 @@ public function get_url( $params = [] ) {
 		if ( isset( $this->resize->get()['type'] ) ) {
 			$path_parts[] = $this->resize->toString();
 		}
-
+		// format is not a must that's what we are changing
```
Comment (selul) [inline inc/image.php]: not sure what means this comment here ?

Comment (GrigoreMihai) [reply inc/image.php]: some old comment from when i was modifying code to remember what i do, it means nothing for others, i will delete it

File: inc/tag_replacer.php (hunk 7 above)
Comment (GrigoreMihai) [reply inc/tag_replacer.php]: it was a link in the video tag, it looks like it got removed from there i think we should put it back there

File: inc/tag_replacer.php [hunk 9]
```diff
@@ -62,22 +62,98 @@ public function init() {
...
 
-			$src = $tmp = $is_slashed ? stripslashes( $images['img_url'][ $index ] ) : $images['img_url'][ $index ];
+			$src = $tmp = $is_slashed ? stripslashes( $images['img_url'][ $index ] ) : $images['img_url'][ $index ]; // remove \
 
 			if ( strpos( $src, $this->upload_resource['content_path'] ) === 0 ) {
-				$src = $tmp = untrailingslashit( $this->upload_resource['content_host'] ) . $src;
+				$src = $tmp = untrailingslashit( $this->upload_resource['content_host'] ) . $src;   // remove last /
 
-				$new_src                     = $is_slashed ? addcslashes( $src, '/' ) : $src;
+				$new_src                     = $is_slashed ? addcslashes( $src, '/' ) : $src;    // make //
```
Comment (preda-bogdan) [inline inc/tag_replacer.php]: Can we remove this comments also?

File: optimole-wp.php [hunk 10]
```diff
@@ -84,7 +84,8 @@ function optml() {
 	}
 
 	if ( ! defined( 'OPTML_DEBUG' ) ) {
-		define( 'OPTML_DEBUG', ( defined( 'TEST_GROUND' ) && TEST_GROUND ) ? true : false );
+		// define( 'OPTML_DEBUG', ( defined( 'TEST_GROUND' ) && TEST_GROUND ) ? true : false );
```
Comment (preda-bogdan) [inline optimole-wp.php]: We should revert to how it was before so we should uncomment this line and remove the one bellow

File: tests/test-generic.php [hunk 11]
```diff
@@ -25,6 +25,6 @@ function test_version_exists() {
 	 */
 	function test_debug_mode() {
 
-		$this->assertFalse( OPTML_DEBUG );
+		$this->assertTrue( OPTML_DEBUG );
```
Comment (preda-bogdan) [inline tests/test-generic.php]: Also this should be as before it should be assert false, the tests were failing from the previous comment where we have explicitly `define( 'OPTML_DEBUG', true );` if we revert to the way it was for production it should work as expected.

File: inc/tag_replacer.php (hunk 9 above)
Comment (GrigoreMihai) [reply inc/tag_replacer.php]: yes, sure

**Comment by preda-bogdan** [comment]
> @preda-bogdan that regex modification seems to solve the issue

@GrigoreMihai 
I would revert the changes and make a new branch from master, just for this changes so we can treat them separately, also this introduces conflicts with the base branch.

File: inc/tag_replacer.php [hunk 12]
```diff
@@ -62,20 +62,96 @@ public function init() {
...
+
 		foreach ( $images[0] as $index => $tag ) {
 			$width     = $height = false;
 			$crop = null;
 			$image_tag = $images['img_tag'][ $index ];
 
 			$is_slashed = strpos( $images['img_url'][ $index ], '\/' ) !== false;
 
-			$src = $tmp = $is_slashed ? stripslashes( $images['img_url'][ $index ] ) : $images['img_url'][ $index ];
+			$src = $tmp = $is_slashed ? stripslashes( $images['img_url'][ $index ] ) : $images['img_url'][ $index ]; // remove \
```
Comment (preda-bogdan) [inline inc/tag_replacer.php]: Remove unnecessary comment

File: inc/tag_replacer.php [hunk 13]
```diff
@@ -62,20 +62,96 @@ public function init() {
...
 
 			if ( strpos( $src, $this->upload_resource['content_path'] ) === 0 ) {
-				$src = $tmp = untrailingslashit( $this->upload_resource['content_host'] ) . $src;
+				$src = $tmp = untrailingslashit( $this->upload_resource['content_host'] ) . $src;   // remove last /
```
Comment (preda-bogdan) [inline inc/tag_replacer.php]: Remove unnecessary comment

**Pull request merged**